test/e2e/version.go

```diff
@@ -0,0 +1,227 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"reflect"
+
+	"github.com/pborman/uuid"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	apicommon "github.com/kubernetes-sigs/federation-v2/pkg/apis/core/common"
+	fedv1a1 "github.com/kubernetes-sigs/federation-v2/pkg/apis/core/v1alpha1"
+	fedclientset "github.com/kubernetes-sigs/federation-v2/pkg/client/clientset/versioned"
+	"github.com/kubernetes-sigs/federation-v2/pkg/controller/sync/version"
+	"github.com/kubernetes-sigs/federation-v2/pkg/controller/util"
+	"github.com/kubernetes-sigs/federation-v2/test/common"
+	testcommon "github.com/kubernetes-sigs/federation-v2/test/common"
+	"github.com/kubernetes-sigs/federation-v2/test/e2e/framework"
+
+	. "github.com/onsi/ginkgo"
+)
+
+var _ = Describe("VersionManager", func() {
+	userAgent := "version-manager"
+
+	f := framework.NewFederationFramework(userAgent)
+
+	tl := framework.NewE2ELogger()
+
+	typeConfig := testcommon.TypeConfigsOrDie(tl)[0]
+
+	targetAPIResource := typeConfig.GetTarget()
+
+	var namespace string
+	var template, override *unstructured.Unstructured
+	var versionName string
+	var versionManager version.VersionManager
+	var expectedStatus fedv1a1.PropagatedVersionStatus
+	var clusterNames []string
+	var versionMap map[string]string
+	var client fedclientset.Interface
+	var qualifiedName util.QualifiedName
+
+	var stopChan chan struct{}
+
+	Context("NamespacedVersionManager", func() {
+		BeforeEach(func() {
+			namespace = f.TestNamespaceName()
+
+			clusterNames := []string{"cluster1", "cluster2", "cluster3"}
+
+			var err error
+			template, _, override, err = testcommon.NewTestObjects(typeConfig, namespace, clusterNames)
+			if err != nil {
+				tl.Fatalf("Failed to retrieve test objects for %q: %v", targetAPIResource.Name, err)
+			}
+			template.SetName(uuid.New())
+			template.SetResourceVersion("templateVer")
+			override.SetResourceVersion("overrideVer")
+
+			versionName = apicommon.PropagatedVersionName(targetAPIResource.Kind, template.GetName())
+			qualifiedName = util.QualifiedName{Namespace: namespace, Name: versionName}
+
+			versionMap = make(map[string]string)
+			for _, name := range clusterNames {
+				versionMap[name] = name
+			}
+
+			expectedStatus = fedv1a1.PropagatedVersionStatus{
+				TemplateVersion: template.GetResourceVersion(),
+				OverrideVersion: override.GetResourceVersion(),
+				ClusterVersions: version.VersionMapToClusterVersions(versionMap),
+			}
+
+			client = f.FedClient(userAgent)
+			versionManager = version.NewNamespacedVersionManager(client, namespace, typeConfig)
+			stopChan = make(chan struct{})
+			versionManager.Sync(stopChan)
```

Comment:
Consider adding a comment that this syncs immediately as there are no API objects.